impl/sql/mutationstorage/queue.go

```diff
@@ -80,27 +126,37 @@ func (m *Mutations) send(ctx context.Context, domainID string, mData []byte, ts
 	return tx.Commit()
 }
 
-// HighWatermark returns the highest timestamp in the mutations table.
-func (m *Mutations) HighWatermark(ctx context.Context, domainID string) (int64, error) {
-	var watermark int64
-	if err := m.db.QueryRowContext(ctx,
-		`SELECT Time FROM Queue WHERE DomainID = ? ORDER BY Time DESC LIMIT 1;`,
-		domainID).Scan(&watermark); err == sql.ErrNoRows {
-		return 0, nil
-	} else if err != nil {
-		return 0, err
+// HighWatermarks returns the highest timestamp in the mutations table.
+func (m *Mutations) HighWatermarks(ctx context.Context, domainID string) (map[int64]int64, error) {
+	watermarks := make(map[int64]int64)
+	rows, err := m.db.QueryContext(ctx,
+		`SELECT ShardID, Max(Time) FROM Queue WHERE DomainID = ? GROUP BY ShardID;`,
```

Comment:
One of the functions above uses `COALESCE` to cover `Time == NULL` case. Is it needed here as well, or not needed above?

Reply:
This one doesn't need it because we're grouping by `ShardID` and therefore don't have a case where `Max` will operate on no rows. -- If there are no rows for a shard, it just won't be returned at all. 

I test for this case in `TestHighwatermarks`

Reply:
Ack.